### PR TITLE
Introduce CMake as build system

### LINKED ISSUE #1029: "make install" target
I've been forced to self-compile Mumble client as the Mumble available in my distro repos is too old to support opus and we also need JACK support, which Mumble doesn't officially support yet.

The prob is that 'sudo make install' fails to install the mumble binary and libs into their correct paths under KXStudio amd64, which is based on Ubuntu 12.04.

Also, there is no mention made of how to install from source within the Mumble INSTALL doc, so I'm just guessing that I'm supposed to run 'make install' as root? I can see the binary and libs etc in the release folder, that much I know.

Here's what happens when I 'sudo make install' Mumble 1.2.4 after its finished compiling:

```
~/src/mumble-1.2.4$ sudo make install
[sudo] password for dan: 
cd celt-0.7.0-build/ && make -f Makefile install
make[1]: Entering directory `/home/dan/src/mumble-1.2.4/celt-0.7.0-build'
make -f Makefile.Release install
make[2]: Entering directory`/home/dan/src/mumble-1.2.4/celt-0.7.0-build'
make[2]: Nothing to be done for `install'.
make[2]: Leaving directory`/home/dan/src/mumble-1.2.4/celt-0.7.0-build'
make[1]: Leaving directory `/home/dan/src/mumble-1.2.4/celt-0.7.0-build'
cd celt-0.11.0-build/ && make -f Makefile install
make[1]: Entering directory`/home/dan/src/mumble-1.2.4/celt-0.11.0-build'
make -f Makefile.Release install
make[2]: Entering directory `/home/dan/src/mumble-1.2.4/celt-0.11.0-build'
make[2]: Nothing to be done for`install'.
make[2]: Leaving directory `/home/dan/src/mumble-1.2.4/celt-0.11.0-build'
make[1]: Leaving directory`/home/dan/src/mumble-1.2.4/celt-0.11.0-build'
cd opus-build/ && make -f Makefile install
make[1]: Entering directory `/home/dan/src/mumble-1.2.4/opus-build'
make -f Makefile.Release install
make[2]: Entering directory`/home/dan/src/mumble-1.2.4/opus-build'
make[2]: Nothing to be done for `install'.
make[2]: Leaving directory`/home/dan/src/mumble-1.2.4/opus-build'
make[1]: Leaving directory `/home/dan/src/mumble-1.2.4/opus-build'
cd src/mumble/ && make -f Makefile install
make[1]: Entering directory`/home/dan/src/mumble-1.2.4/src/mumble'
make -f Makefile.Release install
make[2]: Entering directory `/home/dan/src/mumble-1.2.4/src/mumble'
make[2]: Nothing to be done for`install'.
make[2]: Leaving directory `/home/dan/src/mumble-1.2.4/src/mumble'
make[1]: Leaving directory`/home/dan/src/mumble-1.2.4/src/mumble'
cd plugins/ && make -f Makefile install
make[1]: Entering directory `/home/dan/src/mumble-1.2.4/plugins'
cd link/ && make -f Makefile install                                                                                                                          
make[2]: Entering directory`/home/dan/src/mumble-1.2.4/plugins/link'  
make -f Makefile.Release install  
make[3]: Entering directory `/home/dan/src/mumble-1.2.4/plugins/link'                                                                                         
make[3]: Nothing to be done for`install'.
make[3]: Leaving directory `/home/dan/src/mumble-1.2.4/plugins/link'
make[2]: Leaving directory`/home/dan/src/mumble-1.2.4/plugins/link'
cd manual/ && make -f Makefile install
make[2]: Entering directory `/home/dan/src/mumble-1.2.4/plugins/manual'
make -f Makefile.Release install
make[3]: Entering directory`/home/dan/src/mumble-1.2.4/plugins/manual'
make[3]: Nothing to be done for `install'.
make[3]: Leaving directory`/home/dan/src/mumble-1.2.4/plugins/manual'
make[2]: Leaving directory `/home/dan/src/mumble-1.2.4/plugins/manual'
make[1]: Leaving directory`/home/dan/src/mumble-1.2.4/plugins'
cd overlay_gl/ && make -f Makefile install
make[1]: Entering directory `/home/dan/src/mumble-1.2.4/overlay_gl'
make -f Makefile.Release install
make[2]: Entering directory`/home/dan/src/mumble-1.2.4/overlay_gl'
make[2]: Nothing to be done for `install'.
make[2]: Leaving directory`/home/dan/src/mumble-1.2.4/overlay_gl'
make[1]: Leaving directory `/home/dan/src/mumble-1.2.4/overlay_gl'
```

_This ticket has been migrated from sourceforge. It is thus missing some details like original creator etc.
The original is at https://sourceforge.net/p/mumble/bugs/1029/ ._

### LINKED ISSUE #3808: Improve mumble accessibility: Upgrade qt 5.12 LTS
hi davidebeatrici
I checked mumble's accessibility carefully and found that mumble is not friendly to screen readers. Most of mumble's edit boxes cannot interact with the contents of the edit box using screen readers.
Based on the above question, can the official upgrade the mumble version of qt to 5.12 LTS?
Written in the official qt log, 5.12 LTS rewrites the accessibility module
The following is the update log page for the official qt 5.12 LTS:
[New Features in Qt 5.12](https://wiki.qt.io/New_Features_in_Qt_5.12)
Qt GUI
Added support to Windows UI Automation to the WinRT QPA, allowing Qt-based UWP applications to operate with accessibility and programmatic UI control tools.
thanks

### LINKED ISSUE #3847: PCH header generation fails with MinGW
After switching from Ubuntu Xenial (16.04) to Ubuntu Bionic (18.04) on Travis CI (#3850), our MinGW builds failed with the following error:

```
cc1plus: error: one or more PCH files were found, but they were invalid
cc1plus: error: use -Winvalid-pch for more information
cc1plus: fatal error: release/mumble_pch.hpp: No such file or directory
compilation terminated.
```

We disabled PCH for the builds, but we should investigate what is causing the issue.

### LINKED ISSUE #3996: Migrate project to CMake
WIP https://github.com/davidebeatrici/mumble/tree/cmake

### LINKED ISSUE #4218: Caching the build environment on Azure Pipelines
It looks like our Azure CI/CD always fails when fetching the build environment. If it doesn't fail, still, it would take some time to download the environment before it can start to build.

In #3575, caching has already been discussed and today I came across https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops and learned that Azure started to support caching recently. Maybe we can adopt it?

### PR DESCRIPTION
**Pull request by davidebeatrici**
After 1+ year under development, it's finally here!

Huge thanks to @ZeroAbility, who worked with me since the very beginning and officially joined the team shortly after.

Thanks to @Krzmbrzl for suggestions, testing and asking me everyday to create the pull request, when the project only missed a few bits (which took a few weeks...).

Also thanks to @Kissaki, @TerryGeng, @Septarius and the other testers in `#mumble` and `#mumble-dev` (`@freenode`).

---

Built environments are available at https://dl.mumble.info/build/vcpkg, feel free to download and use them!

We will write a detailed guide on how to use them soon, in the mean time you can take a look at the CI scripts.

---

Advantages over our current qmake project:

- The build directory is not enforced anymore. It was in our qmake project due to some paths being hardcoded (e.g. `RC_ICON` in `.rc` files).
- Translation files are automatically updated, no need to run `updatetranslations.sh` anymore.
The behavior can be disabled by specifying `-Dtranslations=OFF`, useful to minimize the amount of writes for each build.
- The project can (finally) be easily built on Windows, also thanks to vcpkg.

---

Known issues:

- `clean` step deletes `.ts` (translation) files: [QTBUG-76410](https://bugreports.qt.io/browse/QTBUG-76410)
- No option to disable the desktop shortcuts creation. The feature is work in progress and will be finished in the future by @ZeroAbility.
- PCH headers are not used anymore. CMake introduced support for them very recently: https://onqtam.com/programming/2019-12-20-pch-unity-cmake-3-16
The build process is still faster than with our qmake project, but we should consider using PCH headers anyway in future.

---

Fixes #3996 
Closes #3847 (no PCHs are used anymore)
Fixes #3808
Fixes #4218 
Fixes #1029

### CODE REVIEW

**Comment by davidebeatrici** [comment]
While rebasing some changes accidentally went into the wrong commit(s), I will fix it.

**Comment by Kissaki** [comment]
You did not mention anything still being open. So is this the state you suggest to merge as-is?

File: macx/scripts/osxdist.py [hunk 1]
```diff
@@ -325,37 +324,37 @@ def package_server():
...
+	shutil.copy(os.path.join(options.binary_dir, 'README.static.osx'), os.path.join(destdir, 'README'))
+	shutil.copy(os.path.join(options.binary_dir, 'CHANGES'), os.path.join(destdir, 'CHANGES'))
+	shutil.copy(os.path.join(options.binary_dir, 'scripts/murmur.pl'), os.path.join(destdir, 'murmur.pl'))
+	shutil.copy(os.path.join(options.binary_dir, 'scripts/weblist.pl'), os.path.join(destdir, 'weblist.pl'))
+	shutil.copy(os.path.join(options.binary_dir, 'scripts/icedemo.php'), os.path.join(destdir, 'icedemo.php'))
+	shutil.copy(os.path.join(options.binary_dir, 'scripts/weblist.php'), os.path.join(destdir, 'weblist.php'))
+	shutil.copy(os.path.join(options.binary_dir, 'scripts/murmur.ini.osx'), os.path.join(destdir, 'murmur.ini'))
+	shutil.copy(os.path.join(options.binary_dir, 'src/murmur/Murmur.ice'), os.path.join(destdir, 'Murmur.ice'))
+
+	shutil.copy(os.path.join(os.path.join(options.binary_dir, 'murmurd'), os.path.join(destdir, 'murmurd')))
```
Comment (Krzmbrzl) [inline macx/scripts/osxdist.py:347]:
```suggestion
	shutil.copy(os.path.join(options.binary_dir, 'murmurd'), os.path.join(destdir, 'murmurd'))
```
I think you used an additional `os.path.copy` here that accidentally wrapped the whole argument (which is why LGTM complained)

**Comment by TerryGeng** [comment]
Great progress! I can't wait to see this system to be deployed to our CI. :D

**Comment by Krzmbrzl** [comment]
Is environment caching on Azure already turned on (#4218)?
And does this contain a `make install` target (#1029)?

**Comment by Avatat** [comment]
Great job!

**Comment by davidebeatrici** [comment]
Thank you all!

@Krzmbrzl
Caching is turned on, the next CI build should restore the cached environment instead of downloading it.

The `install` target is implemented for all platforms. The Windows build uses it for WiX.

File: macx/scripts/osxdist.py (hunk 1 above)
Comment (davidebeatrici) [reply macx/scripts/osxdist.py]: Thank you, fixed.

**Comment by davidebeatrici** [comment]
@probonopd Now we use the `install` target, no more manual file copy!

**Comment by TerryGeng** [comment]
As discussed in the IRC channel, it seems that some extra steps are needed to run this PR on macOS 10.15.

The first problem comes from `src/tests/CMakeLists.txt`. I changed the condition when adding subdirectories to ignore `CMakeFiles`.
```patch
diff --git a/src/tests/CMakeLists.txt b/src/tests/CMakeLists.txt
index bdfd0690..66c0f4e9 100644
--- a/src/tests/CMakeLists.txt
+++ b/src/tests/CMakeLists.txt
@@ -8,7 +8,7 @@ find_pkg(Qt5 COMPONENTS Test REQUIRED)
 file(GLOB ITEMS RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} "${CMAKE_CURRENT_SOURCE_DIR}/*")

 foreach(ITEM ${ITEMS})
-       if(IS_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/${ITEM}")
+       if(IS_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/${ITEM}" AND NOT ("${ITEM}" STREQUAL "CMakeFiles"))
                add_subdirectory(${ITEM})
                set_target_properties(${ITEM} PROPERTIES RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/tests")
        endif()
```

The other one is about the OpenSSL libraries comes with macOS 10.15. Looks like due to some security reason, now we can't link our build against it. I have to make a longer patch from https://stackoverflow.com/questions/58446253/xcode-11-ld-error-your-binary-is-not-an-allowed-client-of-usr-lib-libcrypto-dy

https://gist.github.com/TerryGeng/eadb589611732a6622b40199448dd5e2

This patch hardcoded the path of homebrew's OpenSSL. I tried to set `PKG_CONFIG_PATH` both in the terminal and by `set(ENV{PKG_CONFIG_PATH} "$ENV{PKG_CONFIG_PATH}:.....")` but it just doesn't work.

**Comment by davidebeatrici** [comment]
The first problem you're encountering is due to the fact that you're running CMake in the repository's root rather than in a dedicated directory (e.g. `Build`). This pollutes the repository with temporary files (which is exactly what happens with our qmake project).

The second problem is due to Apple shipping their own crypto library in recent macOS versions.
You can use OpenSSL from Homebrew by specifying the path to its root folder:
```
cmake -DOPENSSL_ROOT_DIR=/usr/local/opt/openssl ..
```

**Comment by Kissaki** [comment]
Win msvc build with prepared static env.

* [ ] from VS, release mumble.exe works but murmur.exe launches a console window and does not minimize to tray as expected; it seems to not launch correctly
* [ ] cli cmake generate fails for generators default, ninja and nmake; fails to find sndfile
* [ ] from VS debug builds launch with an immediate assertion error on both mumble.exe and murmur.exe

**Comment by davidebeatrici** [comment]
- Confirmed, I have to investigate.
- The build script for Azure Pipelines uses the Ninja generator. Make sure you're specifying `-Dstatic=ON` (the library target's name changes based on static/dynamic).
- Confirmed. It seems to happen when assigning a value to a `QString`. The exception is thrown at:
https://github.com/qt/qtbase/blob/5.12.8/src/corelib/tools/qarraydata.cpp#L167

**Comment by Kissaki** [comment]
I am using static on. I am using the same parameters I added in VS.

**Comment by davidebeatrici** [comment]
Could you paste the parameters, please?

**Comment by Kissaki** [comment]
```batch
"C:\Program Files\CMake\bin\cmake.exe" ^
  -G "Ninja" ^
  "-DVCPKG_TARGET_TRIPLET=x64-windows-static-md" ^
  "-Dstatic=ON" ^
  "-DCMAKE_TOOLCHAIN_FILE=C:/dev/mumble-ecosystem/mumble/buildenv/win64-static-1.4.x-2020-05-27-ecb3c64-1151/scripts/buildsystems/vcpkg.cmake" ^
  "-DIce_HOME=C:\dev\mumble-ecosystem\mumble\buildenv\win64-static-1.4.x-2020-05-27-ecb3c64-1151\installed\x64-windows-static-md" ^
  ..
```

**Comment by Krzmbrzl** [comment]
On my Linux machine when I run
```
mkdir build && cd build
cmake ..
```
The cmake process spits out
```
-- Opus library version: 0.8.0
-- Found Git: /usr/bin/git (found version "2.17.1") 
fatal: No names found, cannot describe anything.
-- Opus package version: 0
-- Opus project version: 0
```

According to https://stackoverflow.com/questions/4916492/git-describe-fails-with-fatal-no-names-found-cannot-describe-anything this is due to `git describe` failing to find the relevant tags.

And if I remember correctly, this is because we are using a custom Opus fork that doesn't contain them. However having this error message occur is not optimal as folks might start wondering what they are doing wrong.
Aren't we setting the Opus version manually anyways? :thinking:
And why do we need a fork in the first place? Are we applying some patches?

**Comment by Krzmbrzl** [comment]
And I'm getting 
```
[ 15%] Built target speex
In file included from /usr/include/dlfcn.h:22:0,
                 from /home/robert/Documents/Git/mumble/overlay_gl/overlay.c:9:
/usr/include/features.h:424:12: fatal error: sys/cdefs.h: No such file or directory
 #  include <sys/cdefs.h>
```
even though `cmake` was happy with everything it found. I guess there's a missing dependency check?

**Comment by davidebeatrici** [comment]
For the Opus issue: `git submodule update --init --recursive`.

For the warnings: yes, I agree.

For the missing header: install `g++-multilib`. I should probably add an option to toggle the overlay cross-compiled target (32 bit library on 64 bit host, for 32 bit games).

**Comment by Krzmbrzl** [comment]
> For the missing header: install g++-multilib. I should probably add an option to toggle the overlay cross-compiled target (32 bit library on 64 bit host, for 32 bit games).

Sounds good. And if it's enabled it'd be nice to check for multilib to be installed on the system during cmake's configuring, so we can print out a nice error message :)

**Comment by ZeroAbility** [comment]
Issues I ran into with Debian 4.19.118-2+deb10u1:

- `slice2cpp` binary does not have execute bit set, required chmod +x in `<vcpkg_root>/installed/x64-linux/tools/Ice`
- `gcc-multilib` install for overlay crosscompile
- `libcap2-bin` install, possibly `libcap-dev` as well
- `libpulse-dev` install for `pulse=ON`
- option `speech-dispatcher=ON` leads to the library not found, `libspeechd-dev` install does not fix but disabling the option does

The final error I have was trying to run `mumble`:

```
./mumble 
qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Aborted
```

**Comment by davidebeatrici** [comment]
- It has to be fixed in the `zeroc-ice` portfile.
- I will add a message.
- `libcap-dev` is required.
- `libpulse-dev` is required until we revamp our code so that it loads PulseAudio dynamically.
- On Debian Bullseye: `speech-dispatcher found | Version: 0.9.1`. Are you sure the package is fully installed?
- The `murmur` target is linked to `Ice::IceUtil` if it's found.
- The Debian package doesn't provide a CMake module, thus we ship one in the `cmake` directory.
Apparently it doesn't work with a vcpkg environment.

As for the missing `xcb` plugin: microsoft/vcpkg#11727

**Comment by davidebeatrici** [comment]
Warnings are now disabled for all 3rdparty targets, except Opus because it's actively maintained. It currently causes 2 warnings:

```cpp
[165/643] Building C object src/mumble/opus/CMakeFiles/opus.dir/celt/x86/x86cpu.c.o
../3rdparty/opus-src/celt/x86/x86cpu.c: In function ‘cpuid’:
../3rdparty/opus-src/celt/x86/x86cpu.c:58:32: warning: unused parameter ‘CPUInfo’ [-Wunused-parameter]
   58 | static void cpuid(unsigned int CPUInfo[4], unsigned int InfoType)
      |                   ~~~~~~~~~~~~~^~~~~~~~~~
../3rdparty/opus-src/celt/x86/x86cpu.c:58:57: warning: unused parameter ‘InfoType’ [-Wunused-parameter]
   58 | static void cpuid(unsigned int CPUInfo[4], unsigned int InfoType)
      |                                            ~~~~~~~~~~~~~^~~~~~~~
```

There is now a `overlay-xcompile` option to toggle the 32 bit library build on Linux, available if the CPU is x86 and 64 bit.

If the option is `ON` (it is by default) the following message is printed:
```
If the 32 bit overlay library fails to compile, make sure the requirements are installed ("g++-multilib" package on Debian-based distributions).
```

As for the Murmur issues on Windows: the command prompt window doesn't show up anymore, I had to pass `WIN32` to `add_executable()`.

Unfortunately I didn't find the cause for the tray icon not showing up yet.

**Comment by Krzmbrzl** [comment]
Starting Mumble built via cmake results in the additional messages printed to the console:
```
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
```
These aren't there if I start Mumble built with qmake. I'm unsure where exactly the difference is coming from though :thinking:

**Comment by TerryGeng** [comment]
I was trying to build mumble with `-DBUILD_TESTING=OFF` but was then stuck at L485 of CMakeLists.txt with error
```
CMake Error at src/mumble/CMakeLists.txt:486 (set_target_properties):
  set_target_properties Can not find target to add properties to:
  test_opus_decode
```

I think this line should be included in some if-statement and executed only if `-DBUILD_TESTING=ON`.

**Comment by davidebeatrici** [comment]
Fixed, thanks!

**Pull request merged**